### PR TITLE
Add go-to-effective command to conflict dependencies

### PR DESCRIPTION
**Pull request by Melody618**
#672, implemented go-to-effective command on conflict dependencies to expand and focus on the only one effective item.

### CODE REVIEW

File: src/explorer/model/ITreeNode.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { ITreeItem } from "./ITreeItem";
+
+export interface ITreeNode {
+    children: ITreeNode[];
+    parent?: ITreeNode | ITreeItem | undefined;
```
Comment (Eskibear) [inline src/explorer/model/ITreeNode.ts:8]: why requrie `ITreeItem`

File: src/handlers/parseRawDependencyDataHandler.ts [hunk 2]
```diff
@@ -31,7 +32,12 @@ export async function parseRawDependencyDataHandler(project: MavenProject): Prom
     const eol: string = "\r\n";
     const prefix: string = "+- ";
     const [treeNodes, conflictNodes] = await parseTreeNodes(treeContent, eol, indent, prefix, project.pomPath);
+    const dependenciesMenu: DependenciesMenu = <DependenciesMenu> project.getChildren()[2];
```
Comment (Eskibear) [inline src/handlers/parseRawDependencyDataHandler.ts:35]: this is tricky to rely on the order. it's not robust if more menus are added under project node.

File: src/extension.ts [hunk 3]
```diff
@@ -59,7 +61,9 @@ async function doActivate(_operationId: string, context: vscode.ExtensionContext
     await vscode.commands.executeCommand("setContext", "vscode-maven:activated", true);
     // register tree view
     await mavenExplorerProvider.loadProjects();
-    context.subscriptions.push(vscode.window.createTreeView("mavenProjects", { treeDataProvider: mavenExplorerProvider, showCollapseAll: true }));
+    const view = vscode.window.createTreeView("mavenProjects", { treeDataProvider: mavenExplorerProvider, showCollapseAll: true });
+    context.subscriptions.push(view);
+    registerCommand(context, "maven.project.goToEffective", (node?: Dependency) => goToEffectiveHandler(view, node));
```
Comment (Eskibear) [inline src/extension.ts:66]: `maven.dependency.goToEffective` looks better

File: package.json [hunk 4]
```diff
@@ -210,6 +210,11 @@
         "category": "Maven",
         "icon": "$(wrench)"
       },
+      {
+        "command": "maven.project.goToEffective",
```
Comment (Eskibear) [inline package.json:214]: `maven.dependency.goToEffective` looks better

File: package.nls.json [hunk 5]
```diff
@@ -17,6 +17,7 @@
     "contributes.commands.maven.project.showDependencies": "Show dependencies",
     "contributes.commands.maven.project.excludeDependency": "Exclude dependency",
     "contributes.commands.maven.project.setDependencyVersion": "Resolve conflict...",
+    "contributes.commands.maven.project.goToEffective": "Go to effective",
```
Comment (Eskibear) [inline package.nls.json:20]: Go to Effective Dependency

File: package.nls.json [hunk 6]
```diff
@@ -17,6 +17,7 @@
     "contributes.commands.maven.project.showDependencies": "Show dependencies",
     "contributes.commands.maven.project.excludeDependency": "Exclude dependency",
     "contributes.commands.maven.project.setDependencyVersion": "Resolve conflict...",
```
Comment (Eskibear) [inline package.nls.json:19]: Resolve Conflict...

File: package.nls.json [hunk 7]
```diff
@@ -17,6 +17,7 @@
     "contributes.commands.maven.project.showDependencies": "Show dependencies",
     "contributes.commands.maven.project.excludeDependency": "Exclude dependency",
```
Comment (Eskibear) [inline package.nls.json:18]: Exclude Dependency

File: package.nls.json [hunk 8]
```diff
@@ -17,6 +17,7 @@
     "contributes.commands.maven.project.showDependencies": "Show dependencies",
```
Comment (Eskibear) [inline package.nls.json:17]: Show Dependencies

File: src/explorer/model/ITreeItem.ts [hunk 9]
```diff
@@ -11,6 +11,7 @@ export interface ITreeItem {
      * If implemented, it will be triggered to get children items.
      */
     getChildren?(): ITreeItem[] | undefined | Promise<ITreeItem[] | undefined>;
+    getParent?(): ITreeItem | undefined | Promise<ITreeItem | undefined>;
```
Comment (Eskibear) [inline src/explorer/model/ITreeItem.ts:14]: Why do you need access to parent of ITreeItem?

File: src/explorer/model/Dependency.ts [hunk 10]
```diff
@@ -42,6 +49,9 @@ export class Dependency extends TreeNode implements ITreeItem {
     public async getChildren(): Promise<Dependency[] | undefined> {
         return Promise.resolve(<Dependency[]> this.children);
     }
+    public async getParent(): Promise<ITreeItem | undefined> {
+        return  Promise.resolve(this.parent);
```
Comment (Eskibear) [inline src/explorer/model/Dependency.ts:53]: strange here.

File: src/handlers/goToEffectiveHandler.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    const pomPath: string = node.projectPomPath;
+    const project: MavenProject | undefined = mavenExplorerProvider.getMavenProject(pomPath);
+    if (project === undefined) {
+        throw new Error("Fail to find maven projects.");
+    }
+
+    const treeNodes = project.treeNodes;
+    const treeItem: Dependency | undefined = await searchFirstEffective(treeNodes, fullArtifactName);
+    if (treeItem === undefined) {
+        throw new Error("Fail to find dependency.");
```
Comment (Eskibear) [inline src/handlers/goToEffectiveHandler.ts:25]: this error message is inconsistent with below

File: src/handlers/goToEffectiveHandler.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+async function searchFirstEffective(treeNodes: Dependency[], fullArtifactName: string): Promise<Dependency | undefined> {
+    let targetItem: Dependency | undefined;
+    const queue: Queue<Dependency> = new Queue();
+    for (const child of treeNodes) {
+        queue.push(child);
+    }
+    while (queue.empty() === false) {
+        const node: Dependency | undefined = queue.pop();
+        if (node === undefined) {
+            throw new Error("Failed to find Dependency.");
```
Comment (Eskibear) [inline src/handlers/goToEffectiveHandler.ts:40]: this error message is inconsistent with above

**Review by Eskibear — CHANGES_REQUESTED** [review]
overall logic is ok. The problem is, what's the difference between `ITreeItem.getParent()` and `ITreeNode.parent`? I see a mixed usage of them. What are they designed for, respectively?

File: src/explorer/model/ITreeNode.ts (hunk 1 above)
Comment (Melody618) [reply src/explorer/model/ITreeNode.ts:8]: for root dependencies, their parent is `DependenciesMenu`, so add `ITreeItem`

Comment (Eskibear) [reply src/explorer/model/ITreeNode.ts:8]: Can it be `undefined`? Is there any use case where you need to access corresponding `DependenciesMenu`?

File: src/explorer/model/MavenProject.ts [hunk 13]
```diff
@@ -26,6 +26,7 @@ export class MavenProject implements ITreeItem {
     public pomPath: string;
     public _fullDependencyText: string;
     public conflictNodes: Dependency[];
+    public treeNodes: Dependency[];
```
Comment (Eskibear) [inline src/explorer/model/MavenProject.ts:29]: dependencies

Comment (Melody618) [reply src/explorer/model/MavenProject.ts:29]: there is already a get function named dependencies

Comment (Eskibear) [reply src/explorer/model/MavenProject.ts:29]: I'm wondering whether getter `dependencies` was acutally called. And there's a bug in line 94, wrongly reading values from plugins.

The name 'treeNodes' is not informative. Let's name it `dependencyNodes`.

**Comment by Eskibear** [comment]
LGTM now

**Review by Eskibear — APPROVED** [review]

**Pull request merged**